feat(Analytics): New payload for analytics SDK

oidacra opened the pull request:
This PR implements user and session management functionality for DotCMS Analytics integration, enabling proper tracking and identification of users across analytics events for Headless sites (NextJS).

### User ID and Session ID Management

User ID Creation and Handling: Implemented logic to generate and manage unique user identifiers for analytics tracking
Session ID Creation and Handling: Added session management capabilities to track user sessions across page visits and interactions
Persistence Logic: Developed the underlying logic to maintain user and session identifiers consistently throughout the user journey

### New Data Structure

New Shape/Schema: Created a new data shape/structure to be sent to the analytics endpoint, incorporating the user and session identification fields
Endpoint Integration: Updated the analytics endpoint communication to handle the new data structure with user and session information

### Technical Implementation

This implementation ensures that:

- Each user receives a unique identifier that persists across sessions
- Sessions are properly tracked and managed
- All analytics data sent to the server includes proper user and session context
- The new data structure follows the required schema for the analytics endpoint

### New Plugin Created: Identity Plugin

Identity Plugin: A new plugin dedicated to managing user identity and session tracking
- User ID generation and persistence
- Session ID creation and management
- Activity tracking functionality
- Local timezone handling
- Session management with throttling to prevent excessive calls

### Updated Plugin: Enrich Plugin
Enhanced Enrich Plugin: Updated to work with the new identity system
- Page data enrichment
- Device information collection
- UTM parameter extraction and caching
- Integration with the new Identity Plugin for comprehensive payload enrichment
- Performance optimizations with caching for static browser data

### Checklist
- [x] Tests
- [ ] Translations
- [ ] Security Implications Contemplated (add notes if applicable)


### Screenshots
![CleanShot 2025-06-20 at 18 10 31@2x](https://github.com/user-attachments/assets/ce927d67-48d5-4a97-a7aa-a97f9acc39e9)


This PR fixes: #32458

oidacra committed:
feat(Analytics): Introduce Angular best practices and update analytics SDK

### Summary of Changes
- Added a new rule file for Angular best practices, focusing on TypeScript, standalone components, and performance optimizations.
- Updated the analytics SDK to include new dependencies for activity tracking and session management.
- Enhanced the analytics payload structure to support enriched data for page views and custom events.
- Improved session handling with user activity tracking and UTM parameter management.

### Additional Notes
- The new Angular best practices aim to improve code maintainability and performance.
- The analytics SDK updates ensure better tracking and data enrichment for user interactions.

oidacra committed:
feat(Analytics): Refactor analytics SDK to enhance event tracking structure

### Summary of Changes
- Introduced new interfaces for  and  to standardize the structure of analytics event payloads.
- Updated the  function to utilize the new request body types for page view and track events.
- Removed deprecated constants and interfaces related to event types and payloads to streamline the codebase.
- Enhanced documentation for analytics event structures to improve clarity and maintainability.

fmontes commented on .cursor/rules/angular-20.mdc:
Ohhhhh nice! :D

oidacra committed:
feat(Analytics): Enhance content analytics initialization and introduce identity plugin

### Summary of Changes
- Refactored the  function to include validation for  and  in the configuration.
- Introduced a new  to manage user ID generation, session management, and activity tracking.
- Updated utility functions to support the new identity plugin and improved session handling.
- Removed the deprecated  function to streamline the analytics setup process.
- Enhanced documentation for better clarity on the new identity management features.

oidacra committed:
feat(Analytics): Introduce new plugins for enhanced data enrichment and identity management

### Summary of Changes
- Added  to enrich analytics payloads with page, device, and UTM data.
- Introduced  for user ID generation, session management, and activity tracking.
- Refactored existing imports to accommodate new plugin structure.
- Updated  to include new dependencies and maintain existing ones.
- Enhanced utility functions for better data extraction and formatting.

### Additional Notes
These changes improve the overall analytics framework by providing more context and user identification for tracking events.

oidacra commented on core-web/libs/sdk/analytics/package.json:
```diff
@@ -20,15 +20,20 @@
     },
     "homepage": "https://github.com/dotCMS/core/tree/main/core-web/libs/sdk/analytics/README.md",
     "dependencies": {
+        "@analytics/activity-utils": "^0.1.16",
```
double-check this.

oidacra committed:
feat(Analytics): Enhance data enrichment and session management in analytics SDK

### Summary of Changes
- Updated the  to include local timezone information in the identity context.
- Enhanced the  plugin to include local time in event payloads.
- Refactored the  to optimize payload enrichment using existing data and removed deprecated utility functions.
- Introduced caching for static browser data and UTM parameters to improve performance.
- Improved session ID management and activity tracking with throttling to prevent excessive calls.

### Additional Notes
These changes improve the accuracy and efficiency of the analytics framework, providing richer context for tracking events and better user session management.

oidacra committed:
chore(Dependencies): Remove unused analytics utility dependencies from package.json files

### Summary of Changes
- Removed deprecated analytics utility dependencies from core-web and analytics SDK package.json files to streamline the project and reduce bloat.

### Additional Notes
These changes help maintain a cleaner codebase by eliminating unnecessary dependencies that are no longer in use.

oidacra committed:
feat(Analytics): Improve analytics initialization and error handling

### Summary of Changes
- Updated the  to ensure that the analytics instance is only assigned to the global window if successfully initialized, with a warning logged if initialization fails.
- Enhanced the  to handle null analytics instances gracefully, preventing errors during rendering.
- Refactored tests to cover new scenarios for null analytics instances and improved error messaging in the  hook.

### Additional Notes
These changes enhance the robustness of the analytics SDK by ensuring that errors during initialization are handled properly, improving the overall user experience and stability of the application.

oidacra committed:
feat(Analytics): Refactor activity tracking to improve event handling

### Summary of Changes
- Updated the activity tracking configuration to remove unused event types and streamline event handling.
- Introduced separate handling for the 'visibilitychange' event to manage tab visibility more effectively.
- Enhanced unit tests to verify the addition and removal of event listeners for both 'click' and 'visibilitychange' events.

oidacra commented:
### QA Testing Notes - DotCMS Analytics Integration

### Setup Instructions
To test this feature, follow these steps:
**1. Environment Setup**

Use the examples/nextjs folder from the core repository to set up NextJS with DotCMS.
Add the following environment variables to your `.env` file:

**NEXT_PUBLIC_DOTCMS_AUTH_TOKEN** - Required to fetch data
**NEXT_PUBLIC_DOTCMS_SITE_KEY** - This is a new environment variable

**2. Code Changes Required**
File: `examples/nextjs/src/views/Page.js`

Add the analytics configuration:
```
analyticsConfig = {
    siteKey: process.env.NEXT_PUBLIC_DOTCMS_ANALYTICS_SITE_KEY || 'KEY_GENERATED_WITH_ANALYTICS_APP_AT_DOTCMS',
    server: process.env.NEXT_PUBLIC_DOTCMS_HOST || 'http://localhost:8080',
    debug: process.env.NODE_ENV === 'development', // Debug only in development
    autoPageView: true // Automatic page view tracking
};
```

Add the import at the top of the file:

`import { DotContentAnalyticsProvider } from "@dotcms/analytics/react";`

Wrap the component return with the provider:
```
return (
    <DotContentAnalyticsProvider config={analyticsConfig}>
        <div className="flex flex-col gap-6 min-h-screen bg-slate-50">
            {pageAsset?.layout.header && (
                <Header navItems={navigation?.children} />
            )}

            <main className="container m-auto">
                <DotCMSLayoutBody
                    page={pageAsset}
                    components={pageComponents}
                    mode={process.env.NEXT_PUBLIC_DOTCMS_MODE}
                />
            </main>

            {pageAsset?.layout.footer && <Footer {...content} />}
        </div>
    </DotContentAnalyticsProvider>
);
```
### Expected Behavior
This setup enables automatic page view tracking for all child components within the DotContentAnalyticsProvider wrapper.
Testing Notes

- Ensure the analytics site key is properly configured
- Verify that page views are automatically tracked when navigating between pages
- Check that debug mode works correctly in development environment
- Test that the analytics server connection is established properly

oidacra commented on core-web/libs/sdk/analytics/src/lib/dotAnalytics/plugin/identity/dot-analytics.identity.utils.ts:
```diff
@@ -0,0 +1,107 @@
+import { safeSessionStorage } from '../../shared/dot-content-analytics.utils';
+
+// Activity tracking state
+let lastActivityTime = Date.now();
```
This `NodeJS.Timeout` is for server side. All this lib is just for Client side.

fmontes replied:
But still we could have this as an object in the client side.

oidacra commented on core-web/libs/sdk/analytics/src/lib/dotAnalytics/plugin/identity/dot-analytics.identity.plugin.ts:
```diff
@@ -0,0 +1,78 @@
+        loaded: () => {
+            // Set up cleanup on page unload
+            if (typeof window !== 'undefined') {
+                window.addEventListener('beforeunload', cleanupActivityTracking);
```
good. Added.

fmontes replied:
TIL!

oidacra committed:
fix(analytics): clean up debug logs

### Summary of Changes
- Downgraded the analytics package version from 0.0.1-alpha.55 to 0.0.1-alpha.54.
- Removed debug logging from the initialization methods in both the dotAnalytics and dotAnalyticsIdentity plugins to streamline the output and reduce noise during initialization.

oidacra committed:
fix(analytics): update script tag selector for analytics key

KevinDavilaDotCMS commented on core-web/libs/sdk/analytics/tsconfig.lib.json:
```diff
@@ -19,5 +19,5 @@
         "**/*.test.jsx"
     ],
-    "include": ["src/**/*.js", "src/**/*.jsx", "src/**/*.ts", "src/**/*.tsx"]
+    "include": ["src/**/*.js", "src/**/*.jsx", "src/**/*.ts", "src/**/*.tsx", "src/**/*.d.ts"]
```
why added .d.ts files?

oidacra replied:
I need removed.

KevinDavilaDotCMS commented on core-web/libs/sdk/analytics/src/lib/dotAnalytics/dot-content-analytics.ts:
```diff
@@ -1,21 +1,61 @@
+    if (typeof window !== 'undefined') {
+        window.addEventListener('beforeunload', cleanup);
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (window as any).__dotAnalyticsCleanup = cleanup;
```
I think we can re-define the namespace of window to add __dotAnalyticsCleanup and avoid use any here

KevinDavilaDotCMS commented on core-web/libs/sdk/analytics/src/lib/dotAnalytics/shared/dot-content-analytics.model.ts:
```diff
@@ -131,41 +197,48 @@ export interface DotAnalyticsPayload {
-export interface DotAnalytics {
-    pageView: (payload?: Record<string, unknown>) => void;
-    track: (eventName: string, payload?: Record<string, unknown>) => void;
+export interface AnalyticsHookParams {
```
In the other sdks, we use the standard to use DotCMS prefix, i think we must maintain it
You can check all standards of the other sdk [here](https://docs.google.com/document/d/16PvgkZj30gjYT5qErp_21t58xgwIuwW7l1CMp4VTCuw/edit?tab=t.o7v8p35vbifq#heading=h.rg7errwt8wc):

oidacra replied:
updated all the types.

KevinDavilaDotCMS commented on core-web/libs/sdk/analytics/src/lib/dotAnalytics/shared/dot-content-analytics.utils.ts:
```diff
@@ -1,15 +1,340 @@
 
-import { dotAnalyticsEnricherPlugin } from '../plugin/dot-analytics.enricher.plugin';
-import { dotAnalytics } from '../plugin/dot-analytics.plugin';
+// Activity tracking state
+let activityListeners: (() => void)[] = [];
+let lastActivityTime = Date.now();
+let inactivityTimer: NodeJS.Timeout | null = null;
+let isThrottled = false;
```
Why this is not a class?

KevinDavilaDotCMS commented on core-web/libs/sdk/analytics/src/lib/dotAnalytics/plugin/identity/dot-analytics.identity.plugin.ts:
```diff
@@ -0,0 +1,79 @@
+    initializeActivityTracking
+} from '../../shared/dot-content-analytics.utils';
+
+/**
+ * Identity Plugin for DotAnalytics
+ * Handles user ID generation, session management, and activity tracking.
+ * This plugin provides consistent identity context across all analytics events.
+ *
+ * Plugin execution order:
+ * 1. Identity Plugin (this) - Injects context
+ * 2. Enricher Plugin - Adds page/device/utm data
+ * 3. Main Plugin - Sends to server
+ */
```
also add the jsdoc for the "@param"

oidacra commented on core-web/libs/sdk/analytics/src/lib/standalone.ts:
```diff
@@ -5,12 +5,19 @@ import { getDataAnalyticsAttributes } from './dotAnalytics/shared/dot-content-an
+        window[ANALYTICS_WINDOWS_KEY] = analytics;
+    } else {
+        console.warn('DotAnalytics: Failed to initialize analytics instance');
+        window[ANALYTICS_WINDOWS_KEY] = null;
```
This is for IFFE. I think it's fine if it's null. When I'm correcting the Traditional thing, I'll double-check it.

oidacra committed:
refactor(analytics): rename DotAnalytics to DotCMSAnalytics and update related configurations

### Summary of Changes
- Renamed  to  across the SDK for consistency.
- Updated configuration interfaces and types to reflect the new naming convention.
- Removed unused test file patterns from the TypeScript configuration.
- Introduced a new activity tracking manager to enhance user activity monitoring and session management.

oidacra committed:
fix(analytics): update data attributes retrieval in standalone analytics initialization

oidacra committed:
refactor(analytics): streamline cleanup handling and improve type definitions

oidacra committed:
fix(analytics): update event payload structure and UTM parameter handling

oidacra committed:
Merge branch 'main' into issue-32458-new-payload-dotcms-analytics-lib

fmontes commented on core-web/libs/sdk/analytics/src/lib/dotAnalytics/shared/dot-content-analytics.model.ts:
```diff
@@ -26,146 +26,319 @@ export interface DotContentAnalyticsConfig {
+    /** The type of event being tracked */
+    event_type: 'pageview' | 'track';
+    /** Page data associated with the event */
+    page: DotCMSPageData;
```
Isn't that this will be dependant if the `event_type === 'pageview'`?

oidacra replied:
fixed.

oidacra commented on core-web/libs/sdk/analytics/src/lib/react/hook/useContentAnalytics.ts:
```diff
@@ -1,63 +1,95 @@
-        track: (eventName: string, payload: Record<string, unknown> = {}) => {
+    const track = useCallback(
+        (eventName: string, payload: Record<string, unknown> = {}) => {
             if (!isInsideEditor()) {
```
I asked about it.
